File: ObjectiveC/FlexGrid101/FlexGrid101/OnDemandController.m
//
//  OnDemandController.m
//  FlexGrid101
//
//  Copyright © 2015 GrapeCity. All rights reserved.
//

#import "OnDemandController.h"
#import "CustomerData.h"
#import "YouTubeCollectionView.h"
#import "XuniFlexGridDynamicKit/XuniFlexGridDynamicKit.h"
#import <XuniCoreDynamicKit/XuniCoreDynamicKit.h>

@interface OnDemandController ()
@property (weak, nonatomic) IBOutlet FlexGrid *flex;
@property (weak, nonatomic) IBOutlet UITextField *textInput;
@property (weak, nonatomic) IBOutlet UISegmentedControl *sortSelect;

@end


@implementation OnDemandController

- (IBAction)startSearching:(id)sender {
    [self.textInput resignFirstResponder];
    [self updateSearch];
}

- (IBAction)searchModeChanged:(id)sender {
    [self updateSearch];
}

-(void) updateSearch
{
    NSArray* selArray = @[@"relevance",@"date",@"viewCount",@"rating",@"title"];

    YouTubeCollectionView* cv = [[YouTubeCollectionView alloc] initWithQuery:self.textInput.text orderBy:[selArray objectAtIndex:self.sortSelect.selectedSegmentIndex]];
    self.flex.collectionView = cv;
    
    
    [cv loadMoreItems:@(50) completion:^{
    }];
}


- (void)viewDidLoad {
    [super viewDidLoad];
    
    NSMutableDictionary* dic = [[NSMutableDictionary alloc]init];
    
    self.flex.columnHeaderFont = [UIFont boldSystemFontOfSize:self.flex.columnHeaderFont.pointSize];
    self.flex.isReadOnly = true;
    
    self.flex.autoGenerateColumns = false;
    
    GridColumn *thumbnail = [[GridColumn alloc] init];
    thumbnail.binding = @"thumbnail";
    thumbnail.width = 150;
    [self.flex.columns addObject:thumbnail];
    
    GridColumn *title = [[GridColumn alloc] init];
    title.binding = @"title";
    title.wordWrap = true;
    [self.flex.columns addObject:title];
    
    GridColumn *channel = [[GridColumn alloc] init];
    channel.binding = @"channelTitle";
    channel.wordWrap = true;
    [self.flex.columns addObject:channel];
    
    [self.flex.flexGridLoadedRows addHandler:^(XuniEventContainer<XuniEventArgs *> *eventContainer) {
        for(int i = 0; i <self.flex.rows.count; i++)
        {
            GridRow* row = [self.flex.rows objectAtIndex:i];
            row.height = 75;
        }
    } forObject:self];
    
    [self.flex.flexGridFormatItem addHandler:^(XuniEventContainer<GridFormatItemEventArgs *> *eventContainer) {
        
        GridColumn* col = self.flex.columns[eventContainer.eventArgs.col];
        if([col.binding isEqualToString:@"thumbnail"] && eventContainer.eventArgs.panel == self.flex.cells)
        {
            NSString* url = (NSString*)[self.flex getCellDataForRow:eventContainer.eventArgs.row inColumn:eventContainer.eventArgs.col formatted:false];
            NSData* imgdata = [dic objectForKey:url];
            if(imgdata == nil)
            {
                dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
                    NSData* imgdata = [NSData dataWithContentsOfURL:[NSURL URLWithString:url]];
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [dic setObject:imgdata forKey:url];
                        [self.flex invalidate];
                    });
                });
                
                eventContainer.eventArgs.cancel = true;
                return;
            }
            
            UIImage  *image = [UIImage imageWithData:imgdata];
            CGRect   r      = [eventContainer.eventArgs.panel getCellRectForRow:eventContainer.eventArgs.row inColumn:eventContainer.eventArgs.col];
            
            CGSize imageSize = image.size;
            CGSize viewSize = r.size;
            
            float hfactor = imageSize.width / viewSize.width;
            float vfactor = imageSize.height / viewSize.height;
            
            float factor = fmax(hfactor, vfactor);
            
            float newWidth = imageSize.width / factor;
            float newHeight = imageSize.height / factor;
            
            CGRect newRect = CGRectMake(r.origin.x+(r.size.width-newWidth)/2,r.origin.y+(r.size.height-newHeight)/2, newWidth, newHeight);
            
            [image drawInRect:newRect];
            
            eventContainer.eventArgs.cancel = true;
        }

    } forObject:self];
    
   
    [self updateSearch];
}


@end

File: Swift/FlexGrid101/FlexGrid101/OnDemandController.swift
//
//  OnDemandController.h
//  FlexGrid101
//
//  Copyright © 2015 GrapeCity. All rights reserved.
//
import UIKit
import XuniFlexGridDynamicKit

class OnDemandController: UIViewController {

    @IBAction func startSearching(sender: AnyObject) {
        self.textInput.resignFirstResponder()
        self.updateSearch()
    }

    @IBAction func searchModeChanged(sender: AnyObject) {
        self.updateSearch()
    }

    func updateSearch() {
        var selArray: [String] = ["relevance", "date", "viewCount", "rating", "title"]
        let cv: YouTubeCollectionView = YouTubeCollectionView(query: self.textInput.text!, orderBy: selArray[self.sortSelect.selectedSegmentIndex])
        self.flex.collectionView = cv
        cv.loadMoreItems(50, completion: {() -> Void in
        })
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        var dic: [NSObject : AnyObject] = [NSObject : AnyObject]()
        self.flex.columnHeaderFont = UIFont.boldSystemFontOfSize(self.flex.columnHeaderFont.pointSize)
        self.flex.isReadOnly = true
        self.flex.autoGenerateColumns = false
        let thumbnail: GridColumn = GridColumn()
        thumbnail.binding = "thumbnail"
        thumbnail.width = 150
        self.flex.columns.addObject(thumbnail)
        let title: GridColumn = GridColumn()
        title.binding = "title"
        title.wordWrap = true
        self.flex.columns.addObject(title)
        let channel: GridColumn = GridColumn()
        channel.binding = "channelTitle"
        channel.wordWrap = true
        self.flex.columns.addObject(channel)
        
        
        self.flex.flexGridLoadedRows.addHandler({(eventContainer: XuniEventContainer!) -> Void in
            for i:UInt in 0 ..< self.flex.rows.count {
                let row: GridRow = self.flex.rows.objectAtIndex(i) as! GridRow
                row.height = 75
            }
        }, forObject: self)
        self.flex.flexGridFormatItem.addHandler({(eventContainer: XuniEventContainer!) -> Void in
            let eventContainerEventArgs = eventContainer.eventArgs as! GridFormatItemEventArgs
            let col: GridColumn = self.flex.columns.objectAtIndex(UInt(eventContainerEventArgs.col)) as! GridColumn
            if (col.binding == "thumbnail") && eventContainerEventArgs.panel == self.flex.cells {
                let url: String = String(self.flex.getCellDataForRow(eventContainerEventArgs.row, inColumn: eventContainerEventArgs.col, formatted: false))
                let imgdata: NSData! = (dic[url] as! NSData!)
                if imgdata == nil {
                    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), {() -> Void in
                        let imgdata: NSData = NSData(contentsOfURL: NSURL(string: url)!)!
                        dispatch_async(dispatch_get_main_queue(), {() -> Void in
                            dic[url] = imgdata
                            self.flex.invalidate()
                        })
                    })
                    eventContainerEventArgs.cancel = true
                    return
                }
                let image: UIImage = UIImage(data:imgdata!)!
                let r: CGRect = eventContainerEventArgs.panel.getCellRectForRow(eventContainerEventArgs.row, inColumn: eventContainerEventArgs.col)
                let imageSize: CGSize = image.size
                let viewSize: CGSize = r.size
                let hfactor: Float = Float(imageSize.width) / Float(viewSize.width)
                let vfactor: Float = Float(imageSize.height) / Float(viewSize.height)
                let factor: Float = fmax(hfactor, vfactor)
                let newWidth: Float = Float(imageSize.width) / factor
                let newHeight: Float = Float(imageSize.height) / factor
                let newRect: CGRect = CGRectMake(r.origin.x + (r.size.width - CGFloat(newWidth)) / 2, r.origin.y + (r.size.height - CGFloat(newHeight)) / 2, CGFloat(newWidth), CGFloat(newHeight))
                image.drawInRect(newRect)
                eventContainerEventArgs.cancel = true
            }
        }, forObject: self)
        self.updateSearch()
    }

    @IBOutlet weak var flex: FlexGrid!
    @IBOutlet weak var textInput: UITextField!
    @IBOutlet weak var sortSelect: UISegmentedControl!
}
//
//  OnDemandController.m
//  FlexGrid101
//
//  Copyright © 2015 GrapeCity. All rights reserved.
//

import XuniCoreDynamicKit